### ISSUE TITLE
Disable post publishing until requirements are met

### ISSUE THREAD

**Issue by joshlevinson** [fusioneng/publishing-checklist#24]
Love the idea of this plugin! I'm implementing something similar that is currently a js solution; it's job is not only to list the requirements, but also to prevent a post from being published if the requirements are not met (while allowing a draft to be saved). Is this something you'd be interested in adding to this plugin as an option? I realize it may not be a feature applicable to all, and so it doesn't lend itself to a default, but I feel that having it available would add to the utility of this plugin.
I'd be glad to submit a PR if you agree.

**Comment by danielbachhuber** [comment]
We deliberately decided _not_ to block publishing. However, I'd be open to a `required` attribute for tasks. If the task was required, and not completed, then publishing would be disabled.

**Comment by joshlevinson** [comment]
:+1: As far as approach goes, does intercepting the publish button to do a requirements check over ajax before submitting the post form sound good to you?

**Comment by danielbachhuber** [comment]
> does intercepting the publish button to do a requirements check over ajax before submitting the post form sound good to you?

Probably not the best approach. Currently, the checklist is entirely dependent on the saved state of the post, not what's in the user's browser. I think we should continue with this model, which means disabling / enabling the Publish button should happen on page load.

**Comment by joshlevinson** [comment]
Hm. I saw that as a more of a current limitation and not necessarily a feature to maintain.
If it were to only occur on page load:
- All new posts would not meet requirements, and would have their publish button disabled
- All existing draft posts that don't meet requirements, that are changed in-browser to meet reqs, would have to be re-saved as a draft first (to update/trigger page reload) before the publish button is enabled
- All existing published posts that *do* meet reqs, that are changed in-browser to *not* meet requirements, would be publishable (since the publish button on page load was enabled).

Come to think of it though, ajax interception wouldn't even work because the callbacks check the saved state of the post. Intercepting the publish action before it saves serves no purpose, as the state will be the same as when the page loaded.

Is an in-browser solution something you're entirely disinterested in?

**Comment by danielbachhuber** [comment]
> Is an in-browser solution something you're entirely disinterested in?

At this point, yes, largely because of:

> Come to think of it though, ajax interception wouldn't even work because the callbacks check the saved state of the post. Intercepting the publish action before it saves serves no purpose, as the state will be the same as when the page loaded.

I think saving to have the publishing checklist update, and publish button un-disabled, is a reasonable expectation of the user, given WordPress isn't an AJAX-y application.

**Comment by joshlevinson** [comment]
> I think saving to have the publishing checklist update, and publish button un-disabled, is a reasonable expectation of the user, given WordPress isn't an AJAX-y application.

This would only work in scenarios where the post being saved is already a draft. It would not for published posts that are edited in such a way that they fail requirements. (I admit I don't know if this is reason (or common) enough to discard the page load solution).

What I meant by an in-browser solution was to provide an option in req registration for a JS callback akin to the PHP one. I imagine checking in-browser for reqs could get a lot hairier though.

**Comment by danielbachhuber** [comment]
> What I meant by an in-browser solution was to provide an option in req registration for a JS callback akin to the PHP one. I imagine checking in-browser for reqs could get a lot hairier though.

Yeah, I think it will be a messier technical challenge than it's worth.

**Comment by celloexpressions** [comment]
Good work on this plugin so far. I'd also be interested in the ability to make items required/disabling save, but that would definitely require a JS-based approach to the whole issue.

I believe the core editor team is working on making ajax-based saving work, so if they're able to pull that off, reworking this plugin to be JS-based would basically be a requirement. Otherwise, you potentially only see progress in a management context as opposed to a writing context. From a UX perspective, a checklist that dynamically updated as you completed tasks on the edit screen would probably be the best experience (that's actually what I expected this to do until I dug into it).

**Comment by pkevan** [comment]
I've started taking a look at add the required attr to the tasks - taking a similar approach to Edit Flow for disabling the publish button.

A couple of thoughts while I was looking at this:

- Even if a task is required you should be able to override this (setting either filtered or Dashboard setting for example allowing Admins by default to override)
- Ability to turn this hard block off by a Dashboard setting (potentially confusing?)
